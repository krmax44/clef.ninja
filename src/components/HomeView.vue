<template>
	<div class="px-4">
		<h2 class="text-2xl">Select your game mode.</h2>

		<div class="flex justify-around py-8 flex-wrap">
			<div class="card-container" @click="selectGamemode('training')">
				<div class="card">
					<div class="card-inner">
						<SchoolIcon :size="48" />

						<h3>Training Mode</h3>
						<p>Get started reading treble and bass clefs.</p>
					</div>
				</div>
			</div>

			<div class="card-container" @click="selectGamemode('arcade')">
				<div class="card">
					<div class="card-inner">
						<GamepadIcon :size="48" />

						<h3>Arcade Mode</h3>
						<p>
							Hit as many correct notes as possible. You have three lives and 60
							seconds. Go!
						</p>
					</div>
				</div>
			</div>

			<div
				class="card-container"
				@click="$store.commit('stage', 'settingsView')"
			>
				<div class="card">
					<div class="card-inner">
						<SettingsIcon :size="48" />

						<h3>Settings</h3>
						<p>
							Setup your MIDI keyboard.
						</p>
					</div>
				</div>
			</div>
		</div>
	</div>
</template>

<script>
import SchoolIcon from 'vue-material-design-icons/SchoolOutline';
import GamepadIcon from 'vue-material-design-icons/GamepadSquareOutline';
import SettingsIcon from 'vue-material-design-icons/SettingsOutline';

export default {
	components: { SchoolIcon, GamepadIcon, SettingsIcon },
	methods: {
		selectGamemode(gamemode) {
			this.$store.commit('gamemode', gamemode);
			this.$store.commit('stage', 'gamePlay');
		}
	}
};
</script>

<style lang="postcss" scoped>
@import '../assets/styles/card-picker.css';

@screen md {
	.card-container {
		@apply w-1/3;
	}
}
</style>
