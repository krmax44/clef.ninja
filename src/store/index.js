import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

const stages = ['gamemodeSelector', 'keyboardSetup', 'game'];

export default new Vuex.Store({
	state: {
		stage: 'gamemodeSelector',
		gamemode: '',
		keyboard: {
			type: 'virtual',
			midi: undefined
		}
	},
	mutations: {
		gamemode(state, gamemode) {
			state.gamemode = gamemode;
		},
		keyboard(state, keyboard) {
			state.keyboard = keyboard;
		},
		nextStage(state) {
			const currentStage = stages.indexOf(state.stage);
			state.stage = stages[currentStage + 1];
		},
		stage(state, stage) {
			state.stage = stages[stage];
		}
	}
});
