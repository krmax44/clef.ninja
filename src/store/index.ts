import Vue from 'vue';
import Vuex from 'vuex';
import InstrumentPiano from '@/instruments/InstrumentPiano';
import InstrumentGuitar from '@/instruments/InstrumentGuitar';

Vue.use(Vuex);

const instruments = {
	piano: InstrumentPiano,
	guitar: InstrumentGuitar
};

export default new Vuex.Store({
	state: {
		stage: 'homeView',
		gamemode: '',
		keyboard: {
			type: 'virtual',
			midi: undefined
		},
		instrument: InstrumentPiano
	},
	mutations: {
		gamemode(state, gamemode) {
			state.gamemode = gamemode;
		},
		keyboard(state, keyboard) {
			state.keyboard = keyboard;
		},
		stage(state, stage) {
			state.stage = stage;
		},
		instrument(state, instrument: string) {
			if (Object.keys(instruments).includes(instrument)) {
				state.instrument = instruments[instrument as 'guitar' | 'piano'];
			}
		}
	},
	getters: {
		midi(state) {
			return state.keyboard.type === 'midi' && state.keyboard.midi;
		}
	}
});
