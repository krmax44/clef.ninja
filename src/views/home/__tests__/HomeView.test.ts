import { shallowMount, createLocalVue } from '@vue/test-utils';
import HomeView from '../index.vue';

import store from '@/store';
import PortalVue from 'portal-vue';

function createWrapper() {
	const localVue = createLocalVue();
	localVue.use(PortalVue);

	return shallowMount(HomeView, { store, localVue });
}

describe('HomeView', () => {
	const wrapper = createWrapper();

	it('sets the gamemode', () => {
		wrapper.find('.card-container').trigger('click');
		expect((wrapper.vm as any).$store.state.gamemode).toBe('training');
		expect((wrapper.vm as any).$store.state.stage.name).toBe('gamePlay');
	});

	it('renders correctly', () => {
		expect(wrapper.element).toMatchSnapshot();
	});
});
