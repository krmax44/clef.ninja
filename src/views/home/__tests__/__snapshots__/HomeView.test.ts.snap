// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeView renders correctly 1`] = `
<div
  class="px-4"
>
  <portal
    to="header-right-small"
  >
    <div
      class="flex items-center"
    >
      <div
        class="btn-options"
        role="button"
      >
        <settingsicon-stub
          fillcolor="#718096"
          size="32"
          title="Settings Outline icon"
        />
      </div>
    </div>
  </portal>
   
  <h2
    class="text-2xl"
  >
    Select your game mode.
  </h2>
   
  <div
    class="flex justify-around py-8 flex-wrap"
  >
    <div
      class="card-container"
    >
      <div
        class="card"
      >
        <div
          class="card-inner"
        >
          <schoolicon-stub
            fillcolor="currentColor"
            size="48"
            title="School Outline icon"
          />
           
          <h3>
            Training Mode
          </h3>
           
          <p>
            Get started reading treble and bass clefs.
          </p>
        </div>
      </div>
    </div>
     
    <div
      class="card-container"
    >
      <div
        class="card"
      >
        <div
          class="card-inner"
        >
          <gamepadicon-stub
            fillcolor="currentColor"
            size="48"
            title="Gamepad Square Outline icon"
          />
           
          <h3>
            Arcade Mode
          </h3>
           
          <p>
            
						Hit as many correct notes as possible. You have three lives and 60
						seconds. Go!
					
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
